### PR TITLE
modified ETA - decreased I/O cost

### PR DESCRIPTION
**Pull request by aduriseti**
now uses ciel(num_remaining * global_time_avg_per_item) to estimate ETA

### CODE REVIEW

**Comment by aduriseti** [comment]
Now also prevents I/O cost during progress bar update from occuring more than 1/dt (10 by default) times a second

**Comment by praiskup** [comment]
Have you tried to look at python-progress/python-progress?  That code has
been already (a) fixed and (b) optimized.

**Comment by praiskup** [comment]
Also, have a look at verigak/progress#23.

File: progress/__init__.py [hunk 1]
```diff
@@ -65,19 +65,20 @@ def start(self):
...
+        self._pending += n
+        #avoid performing computationally intensive tasks more than 1/dt (10 by default) times a second
+        if dt > self.time_threshold:
+            self._xput.append((self._pending) / dt)
+            self._ts = now
+            self._pending = 0
+            self.update()
         self.index = self.index + n
-        self.update()
+
```
Comment (verigak) [inline progress/__init__.py]: So you did 3 changes here:
1) Bail out early if n <= 0 (good)
2) Flattened the time_threshold check (I don't see the reason for this, they are 2 distinct cases, and you are changing self._pending just to set it to 0 later)
3) Call update() only when we are passed the threshold (good)

I would keep 1 and 3, but I don't see any gains with 2.

File: progress/__init__.py [hunk 2]
```diff
@@ -95,8 +96,8 @@ def __init__(self, *args, **kwargs):
 
     @property
     def eta(self):
-        avg = self.avg
-        return int(ceil(self.remaining / avg)) if avg else 0
+        if not self.index: return 0
+        return int(ceil((self.max - self.index) * self.elapsed/self.index))
```
Comment (verigak) [inline progress/__init__.py]: Why you are not using `self.remaining` here? It has an extra bounds check. Also, with this change you are basically disabling the moving average algorithm and use plain average.

**Comment by verigak** [comment]
Please see my inline comments. Also make sure you follow PEP8. I'm getting 6 issues with your PR:

```
__init__.py:69:5: E303 too many blank lines (2)
__init__.py:70:18: E701 multiple statements on one line (colon)
__init__.py:74:9: E265 block comment should start with '# '
__init__.py:74:80: E501 line too long (103 > 79 characters)
__init__.py:83:5: E303 too many blank lines (2)
__init__.py:99:26: E701 multiple statements on one line (colon)

```

File: progress/__init__.py (hunk 2 above)
Comment (aduriseti) [inline progress/__init__.py]: There's no reason I'm not using self.remaining.  The reason I disabled the moving average algorithm is that it gives highly variable ETA's for my use case: loops over 1000s of computations with infrequent API calls.  For this - average ETA works better I've found.

File: progress/__init__.py (hunk 1 above)
Comment (aduriseti) [inline progress/__init__.py]: There's no logic to #2 - the motivation is purely stylistic.  I think you are right - there is one extra assignment operation incurred by my method.

File: progress/__init__.py (hunk 2 above)
Comment (aduriseti) [inline progress/__init__.py]: I actually use right now a ETA confidence interval implementation discussed in this stack overflow post: http://stackoverflow.com/questions/933242/smart-progress-bar-eta-computation.  I also found the method described by SPWorley in that post to be better than a simple average.

**Comment by aduriseti** [comment]
Addressed pep8 flags and inline comments

**Comment by verigak** [comment]
The thing is that you basically reverted the moving average algorithm to a simple average as a side-effect of your change. I'm open to alternative algorithms, but make it clear what you are doing. For example now `_xput` is not being used at all, but still stuff are appended to it.

I would rather keep just the `update()` change and if you'd like to suggest a different algorithm do it in a separate PR, making it clear what the change is all about.

**Comment by aduriseti** [comment]
Makes sense - I'll fork and make a separate PR.

**Comment by praiskup** [comment]
@aduriseti have you tried the project I pointed you at?  There is O(1) algorithm implemented ..
I had similar issue with the code as you (high network speed).
Most probably you wouldn't have a problem anymore.  But I just wan''t to be sure that I should
not fix my fork, too.

**Comment by praiskup** [comment]
Ah, I now see clearly your issue.  This is exactly what I told to @verigak ... in verigak/progress#23 and verigak/progress#24, that there is non-constant algorithm for avg() --> iterating through array for each next() call is **wrong**.

Just moving update() is not correct, because we don't calculate the last N into average (I've already told to @verigak).  And during the first calls to next() this makes progress give you _big_ lie usually, because the speed is getting higher until you reach your bandwidth limit.

Also, you must count with extremely slow connections too ... so there could happen that N==0 is returned and the update() gets very slowly (or never) called.  Moving update makes this worse.

I would suggest you implementing the constant algorithm so you don't have to limit the update() calls, or simply have a look at python-progress/python-progress.

**Comment by aduriseti** [comment]
I have checked it out - I like the time based as opposed to item based window size

**Comment by albertferras** [comment]
+1 can we merge this?

**Comment by aduriseti** [comment]
The consensus was that I create separate pull requests for the scheduling and optimization changes.  I will do it - but work has been busy recently.

**Pull request closed without merging**